common/scala/src/main/scala/whisk/common/tracing/OpenTracingProvider.scala

```diff
@@ -0,0 +1,195 @@
+        map.toMap
+      }
+  }
+
+  /**
+   * Get the current TraceContext which can be used for downstream services
+   *
+   * @param transactionId
+   * @return
+   */
+  override def setTraceContext(transactionId: TransactionId, context: Option[Map[String, String]]) = {
+    context.foreach { scalaMap =>
+      val ctx: SpanContext = tracer.extract(Format.Builtin.TEXT_MAP, new TextMapExtractAdapter(scalaMap.asJava))
+      contextMap.put(transactionId.meta.id, ctx)
+    }
+  }
+
+  private def clear(transactionId: TransactionId): Unit = {
+    spanMap.get(transactionId.meta.id).foreach {
+      case head :: Nil =>
+        head.finish()
+        spanMap.remove(transactionId.meta.id)
+        contextMap.remove(transactionId.meta.id)
+      case head :: tail =>
+        head.finish()
+        spanMap.put(transactionId.meta.id, tail)
+      case Nil =>
+    }
+  }
+
+  private def configureCache[T, R](): collection.concurrent.Map[T, R] =
+    Caffeine
+      .newBuilder()
+      .expireAfterAccess(tracingConfig.cacheExpiry.toSeconds, TimeUnit.SECONDS)
+      .build()
+      .asMap()
+      .asScala
+      .asInstanceOf[collection.concurrent.Map[T, R]]
+}
+
+trait WhiskTracer {
+  def startSpan(logMarker: LogMarkerToken, transactionId: TransactionId): Unit = {}
+  def finishSpan(transactionId: TransactionId): Unit = {}
+  def error(transactionId: TransactionId): Unit = {}
+  def getTraceContext(transactionId: TransactionId): Option[Map[String, String]] = None
+  def setTraceContext(transactionId: TransactionId, context: Option[Map[String, String]]): Unit = {}
+}
+
+object WhiskTracerProvider {
+  val tracingConfig = loadConfigOrThrow[TracingConfig](ConfigKeys.tracing)
+
+  val tracer: WhiskTracer = createTracer(tracingConfig)
+
+  private def createTracer(tracingConfig: TracingConfig): WhiskTracer = {
+
+    tracingConfig.zipkin match {
+      case Some(zipkinConfig) => {
+        if (!GlobalTracer.isRegistered) {
+          val sender: Sender = OkHttpSender.create(zipkinConfig.getUrl)
+          val spanReporter = AsyncReporter.create(sender)
+          val braveTracing = Tracing
+            .newBuilder()
+            .localServiceName(tracingConfig.component)
+            .spanReporter(spanReporter)
+            .sampler(Sampler.create(zipkinConfig.sampleRate.toFloat))
+            .build()
+
+          //register with OpenTracing
+          GlobalTracer.register(BraveTracer.create(braveTracing))
+
+          sys.addShutdownHook({ spanReporter.close() })
+        }
+      }
+      case None =>
+    }
+
+    if (GlobalTracer.isRegistered)
+      new OpenTracer(GlobalTracer.get(), tracingConfig)
+    else
+      NoopTracer
+  }
+}
+
+private object NoopTracer extends WhiskTracer
+case class TracingConfig(component: String, cacheExpiry: Duration, zipkin: Option[ZipkinConfig] = None)
+case class ZipkinConfig(url: String, sampleRate: String) {
+  def getUrl = s"$url/api/v2/spans"
```

Comment:
Can we rename this to `generateUrl` or something? It's unclear otherwise what the difference between `url` (rename to `baseUrl`?) and `getUrl` is without looking up the code.